Sparc/MC680x0 Mach-O issues

Rot127 opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository.
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style).
- [x] I've documented every `RZ_API` function and struct this PR changes.
- [ ] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [ ] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

Fixes three things:
1. Adds two more register layouts (Sparc, m68k) for Mach-O thread_status notes.
2. Fixes the issue of endianess not being handled properly for Mach-O files (it assumed only PPC is always big endian or ignored it).
3. Adds a command to switch between the different binaries of a fat binary.

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

Added

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

closes https://github.com/rizinorg/rizin/issues/4359

Rot127 committed:
Add thread_state register layouts for Mac-O Sparc and MC680x0 binaries.

Rot127 committed:
Prevent hitting assert by passing NULL in some cases.

If name is not a valid arch identifier, it should complain later.

Rot127 committed:
Switch to m68k name, to match the asm.arch plugin name.

Rot127 committed:
Don't ignore endianess of fat Mach-O binaries.

And if it isn't ignored check for actual endianess,
not for being a PPC binary.

Rot127 committed:
Add obs command to select the different versions of fat binaries.

Rot127 committed:
clang-format

wargio commented on librz/bin/format/mach0/mach0_specs.h:
```diff
@@ -190,6 +190,38 @@ struct arm_thread_state64 {
+ * I use this in the hope the Mach-O and ELF devs had some
+ * commen documentation they were refering to.
+ */
+struct sparc_thread_state64 {
```
are you sure that `LC_THREAD` is supported by the old mach0 bins?

wargio replied:
if you have a binary for both, i could check via otool if it says anything.

Rot127 replied:
Actually, you are right. I just never thought about that the old Sparc processors might all be one threaded.

Rot127 committed:
Fix import tests of mach-o

Rot127 committed:
Readd filenames

Rot127 committed:
Properly rest xtr_data->loaded

Rot127 committed:
Add type annotation.

notxvilka commented on librz/bin/bfile.c:
```diff
@@ -80,9 +80,26 @@ static RzBinPlugin *get_plugin_from_buffer(RzBin *bin, const char *pluginname, R
+	}
+}
+
+RZ_API bool rz_bin_file_set_xtr_data_as_current_obj(RzBin *bin, RzBinFile *bf, RzBinObjectLoadOptions *opts, RzBinXtrData *data) {
```
Add Doxygen please

notxvilka commented on librz/core/cbin.c:
```diff
@@ -4649,37 +4652,24 @@ static int bin_versioninfo(RzCore *r, PJ *pj, int mode) {
-RZ_API int rz_core_bin_set_arch_bits(RzCore *r, const char *name, const char *arch, ut16 bits) {
-	int fd = rz_io_fd_get_current(r->io);
-	RzIODesc *desc = rz_io_desc_get(r->io, fd);
+RZ_API int rz_core_bin_set_arch_bits(RzCore *r, RZ_DEPRECATE const char *filename, const char *arch, ut16 bits) {
```
Please add Doxygen since you modified it

notxvilka reviewed (approved):
Looks good!

Rot127 committed:
Update doxygen and remove RZ_DEPRECATED flag

Rot127 committed:
Remove NULL check for arch.

Apparently, arch is so commonly NULL by default enforcing RZ_NONNULL might just breaks too much.

Rot127 commented:
Windows test fail:
> Build execution time has reached the maximum allowed time for your plan (90 minutes).

notxvilka committed:
Merge branch 'dev' into sparc-mach-o

notxvilka commented on librz/bin/format/mach0/mach0.c:
```diff
@@ -1070,8 +1070,36 @@ static int parse_thread(struct MACH0_(obj_t) * bin, struct load_command *lc, ut6
+			return false;
+		}
+		if (rz_buf_fread_at(bin->b, ptr_thread,
+			    (ut8 *)&bin->thread_state.mc680x0, bin->big_endian ? "36L" : "36l", 1) == -1) {
```
Is this a typo? Should this one be `thread_state.sparc` instead?

notxvilka reviewed (changes requested):
Seems there is a copy-paste error.

notxvilka committed:
Mach-O: use the correct thread state structure for SPARC
